### ISSUE TITLE
Fine-tuning on handwritten dataset yields unsatisfactory results

### ISSUE THREAD

**Issue by 2ayush2** [datalab-to/surya#430]
I have fine-tuned the model on a handwritten dataset, but the results after fine-tuning have not met expectations. The accuracy and overall performance remain below the desired level.

Dataset: Handwritten text dataset (custom)

Goal: Improve recognition accuracy for handwriting

Observation: No significant improvement after fine-tuning

Has anyone experienced similar results or found effective approaches for improving fine-tuning performance on handwritten data? Any guidance or tips would be appreciated.

**Comment by tarun-menta** [comment]
@2ayush2 If you can share more about the following, I can help:
- Your training setup (GPUs, batch size, #samples, etc)
- What type of images you're aiming to train on
- The kind of errors you see in base surya

As we mentioned, the finetuning script is very bare-bones. If you want to collaborate to get in some data to our main model for the next release, happy to help!

**Comment by 2ayush2** [comment]
This was the code i had tried
from __future__ import annotations
from typing import Optional, Tuple, Any, Dict
import numpy as np
import torch
from datasets import load_dataset, Dataset
from transformers import Trainer, TrainingArguments


from surya.common.surya import SuryaModel
from surya.common.surya.processor import SuryaOCRProcessor
from surya.foundation import FoundationPredictor
from surya.common.surya.processor.schema import ImageInput, TextInput
from surya.common.surya.schema import TaskNames
from surya.common.util import get_top_scripts, SCRIPT_TOKEN_MAPPING

# Constants
OCR_TASK_NAME = TaskNames.ocr_with_boxes
OCR_MAX_IMAGE_SIZE = (1024, 512)

class SuryaOCRDataset(torch.utils.data.Dataset):
    def __init__(self, processor: SuryaOCRProcessor, hf_split: Dataset):
        super().__init__()
        self.hf_dataset = hf_split
        self.processor = processor

    def __len__(self):
        return len(self.hf_dataset)

    def get_script_text(self, text: str) -> str:
        scripts = get_top_scripts(text)
        return "".join(SCRIPT_TOKEN_MAPPING[script] for script in scripts)

    def __getitem__(self, index: int) -> Dict[str, Any]:
        try:
            data = self.hf_dataset[index]
            image = data["image"].convert("RGB")
            image = np.asarray(image, dtype=np.float32)
            image = self.processor.scale_to_fit(image, max_size=OCR_MAX_IMAGE_SIZE)

            gt_text = data["text"]
            gt_text = self.get_script_text(gt_text) + gt_text

            return {
                "task": OCR_TASK_NAME,
                "inputs": [
                    ImageInput(type="image", image=image, rotated=False),
                    TextInput(type="text", text=""),  # Keep empty to match format
                    TextInput(type="text", text=gt_text),
                ],
            }
        except Exception:
            import traceback
            traceback.print_exc()
            return self.__getitem__((index + 1) % self.__len__())

class SuryaOCRDataCollator:
    def __init__(self, processor: SuryaOCRProcessor, max_sequence_length: Optional[int] = None):
        self.processor = processor
        self.max_sequence_length = max_sequence_length

    def __call__(self, inputs):
        processed_batch = self.processor(inputs, padding_side="right")

        if self.max_sequence_length:
            processed_batch["input_ids"] = processed_batch["input_ids"][:, :self.max_sequence_length]
            processed_batch["attention_mask"] = processed_batch["attention_mask"][:, :self.max_sequence_length]
            processed_batch["position_ids"] = processed_batch["position_ids"][:, :self.max_sequence_length]

        lm_labels = processed_batch["input_ids"].clone()
        skip_mask = (
            (lm_labels == self.processor.pad_token_id)
            | (lm_labels == self.processor.bos_token_id[OCR_TASK_NAME])
            | (lm_labels == self.processor.eoi_token_id)
            | (lm_labels == self.processor.image_token_id)
        )
        lm_labels[skip_mask] = -100
        processed_batch["labels"] = lm_labels
        return processed_batch

def load_model_and_processor(checkpoint_path: Optional[str] = None) -> Tuple[SuryaModel, SuryaOCRProcessor]:
    foundation_predictor = FoundationPredictor(checkpoint=checkpoint_path)
    return foundation_predictor.model, foundation_predictor.processor

def train_ocr(
    checkpoint_path: Optional[str] = None,
    dataset_name: str = "Sylk121/Surya-ocr",
    output_dir: str = "./surya_ocr_finetuned",
    num_proc: int = 16,
    val_ratio_if_missing: float = 0.1,
    max_sequence_length: int = 1024,
):
    model, processor = load_model_and_processor(checkpoint_path)
    raw = load_dataset(dataset_name, num_proc=num_proc)
    if "validation" in raw:
        train_split = raw["train"]
        val_split = raw["validation"]
    else:
        split = raw["train"].train_test_split(test_size=val_ratio_if_missing, seed=42)
        train_split, val_split = split["train"], split["test"]

    # Wrap splits
    train_dataset = SuryaOCRDataset(processor, train_split)
    eval_dataset = SuryaOCRDataset(processor, val_split)

    # Collator
    collator = SuryaOCRDataCollator(processor, max_sequence_length=max_sequence_length)

    # Training arguments
    training_args = TrainingArguments(
        output_dir=output_dir,
        per_device_train_batch_size=8,
        per_device_eval_batch_size=8,
        gradient_accumulation_steps=2,
        learning_rate=1e-5,
        num_train_epochs=20,
        gradient_checkpointing=True,
        optim="adafactor",
        logging_dir="./logs",
        logging_steps=60,
        save_total_limit=1,
        dataloader_pin_memory=True,
        dataloader_num_workers=2,
        remove_unused_columns=False,
        eval_strategy="steps",
        eval_steps=50,
    )

    trainer = Trainer(
        model=model,
        args=training_args,
        train_dataset=train_dataset,
        eval_dataset=eval_dataset,
        data_collator=collator,
    )

    print("Training configuration:", training_args)
    trainer.train()
    print("Training finished.")

train_ocr()

This is the dataset I have tried: https://huggingface.co/datasets/Sylk121/Surya-ocr. Could you tell me where I might be going wrong? I’m also experiencing overfitting.

**Comment by tarun-menta** [comment]
@2ayush2 it seems to me that the dataset is a little too small to avoid overfitting. How many epochs/steps are you running for? 

You could also try LORA instead of full finetuning, and train for lesser steps

**Comment by tarun-menta** [comment]
Any updates? Happy to discuss further

**Comment by tarun-menta** [comment]
Closing
